ros_ws/src/autonomous/src/wall_following.cpp

```diff
@@ -0,0 +1,336 @@
+#include <iostream>
+#include <array>
+#include <ros/ros.h>
+#include <ros/console.h>
+#include "sensor_msgs/LaserScan.h"
+#include "std_msgs/Float64.h"
+#include "drive_msgs/drive_param.h"
+#include "drive_msgs/pid_input.h"
+#include <math.h>
+
+
+float g_right_error = 0;
+float g_right_prev_error = 0;
+float g_right_corrected_angle = 0;
+
+float g_left_error = 0;
+float g_left_prev_error = 0;
+float g_left_corrected_angle = 0;
```

Comment:
May put those global variables at least in a `struct` e.g. 
```C++
template<typename T>
struct control_correction_data {
    T error;
    T prev_error;
    T corrected_angle;
};
control_correction_data<float> g_right;
control_correction_data<float> g_left;
```
the `template` would allow to easily change the type to `double` or another type later on.

Also according to a [stackoverflow post](http://web.archive.org/web/20181219233245/https://stackoverflow.com/questions/37502096/ros-binding-a-callback-function-and-a-object-member-to-a-subscriber-node) you should be able to put all the global variables in the class `WallFollowing` without any problems.